src/main/scala/org/globalforestwatch/Analysis.scala

```diff
@@ -1,55 +1,169 @@
 }
 
 object Analysis {
+  val logger = Logger.getLogger(getClass.getName)
+
   val aggByYear = udaf(LossYearAgg)
+  val aggHist = udaf(HistogramAgg)
 
   def addColumns(df: DataFrame, analyses: List[Analysis]): DataFrame = {
     analyses.foldLeft(df){ (result, analysis) =>
-      result.withColumn(analysis.name, analysis.condition)
+      result.withColumn(analysis.name, analysis.content)
     }
   }
 
   def aggColumns(analyses: List[Analysis]): List[Column] = {
     analyses.map(a => a.agg as a.name)
   }
 
-  object TreeCoverLossTotalYearly extends Analysis("tree_cover_loss_total_yearly") {
-    def condition: Column =
+  // ---------------------------------------------------- Pixelwise Aggregations
+  object TreeCoverLossTotalYearlyPW extends Analysis("tree_cover_loss_total_yearly") {
+    def content: Column =
       when(TreeCoverLoss.col > 0 && TreeCoverDensityPercent2000.col > 30, col("area")) as name
-    def agg: Column =
-      Analysis.aggByYear(TreeCoverLoss.col as "loss_year", col(name) as "area") as name
   }
 
-  object TreeCoverLoss90Yearly extends Analysis("tree_cover_loss_tcd90_yearly") {
-    def condition: Column =
+  object TreeCoverLoss90YearlyPW extends Analysis("tree_cover_loss_tcd90_yearly") {
+    def content: Column =
       when(TreeCoverLoss.col > 0 && TreeCoverDensityPercent2000.col > 90, col("area")) as name
-    def agg: Column =
-      Analysis.aggByYear(TreeCoverLoss.col as "loss_year", col(name) as "area") as name
   }
 
-  object TreeCoverLossPrimaryForestYearly extends Analysis("tree_cover_loss_primary_forest_yearly") {
-    def condition: Column =
+  object TreeCoverLossPrimaryForestYearlyPW extends Analysis("tree_cover_loss_primary_forest_yearly") {
+    def content: Column =
       when(TreeCoverLoss.col > 0 && PrimaryForest.col > 0, col("area")) as name
-    def agg: Column =
-      Analysis.aggByYear(TreeCoverLoss.col as "loss_year", col(name) as "area") as name
   }
 
-  object TreeCoverLossPeatYearly extends Analysis("tree_cover_loss_peat_yearly") {
-    def condition: Column =
+  object TreeCoverLossPeatYearlyPW extends Analysis("tree_cover_loss_peat_yearly") {
+    def content: Column =
       when(TreeCoverLoss.col > 0 && GFWProPeatlands.col > 0, col("area")) as name
-    def agg: Column =
-      Analysis.aggByYear(TreeCoverLoss.col as "loss_year", col(name) as "area") as name
   }
 
+
+  // --------------------------------------------------- Rasterized Aggregations
+  class HistogramVisitor(val result: Array[Double], key: Tile) extends GridVisitor[Raster[Tile], Array[Double]] {
+    def visit(data: Raster[Tile], c: Int, r: Int) = {
+      val ix = key.get(c, r)
+      if (ix >= 0)
+        result(ix) += data.tile.getDouble(c, r)
+    }
+  }
+
+  def rasterizeToHistogram(
+    key: SpatialKey,
+    geom: Geometry,
+    yearTile: Tile,
+    criterionTile: Tile
+  ): Array[Double] = {
+    logger.warn(s"Rasterizing for ${key}; tile is ${yearTile.cols}×${yearTile.rows}")
```

Comment:
It is worth noting that this log message never materializes.  Not sure how that can be while running on `master=local`.  Log4j root level seems to be set to INFO.